File: ospylib/hardware/arch/windows/processor.cpp
/*
Copyright (c) 2022, Bamboooz
All rights reserved.
Use of this source code is governed by a BSD-style license that can be
found in the LICENSE file.
*/

#include <cpuid.h>
#include <stdio.h>
#include <string>
#include <stdlib.h>
#include <string.h>
#include <iostream>
#include <windows.h>
#include <unordered_map>

using namespace std;

static const unordered_map<string, string> vendor_map = {
    { "AMDisbetter!", "AMD" },
    { "AuthenticAMD", "AMD" },
    { "CentaurHauls", "IDT" },
    { "CyrixInstead", "Cyrix" },
    { "GenuineIntel", "Intel" },
    { "TransmetaCPU", "Transmeta" },
    { "GenuineTMx86", "Transmeta" },
    { "Geode by NSC", "National Semiconductor" },
    { "NexGenDriven", "NexGen" },
    { "RiseRiseRise", "Rise" },
    { "SiS SiS SiS", "SiS" },
    { "UMC UMC UMC", "UMC" },
    { "VIA VIA VIA", "VIA" },
    { "Vortex86 SoC", "DM&P Vortex86" },
    { "  Shanghai  ", "Zhaoxin" },
    { "HygonGenuine", "Hygon" },
    { "Genuine  RDC", "RDC Semiconductor Co. Ltd." },
    { "E2K MACHINE", "MCST Elbrus" },
};

const string model() {
    uint32_t cpuid_values[12];

    __get_cpuid(0x80000002, cpuid_values + 0x0, cpuid_values + 0x1, cpuid_values + 0x2, cpuid_values + 0x3);
    __get_cpuid(0x80000003, cpuid_values + 0x4, cpuid_values + 0x5, cpuid_values + 0x6, cpuid_values + 0x7);
    __get_cpuid(0x80000004, cpuid_values + 0x8, cpuid_values + 0x9, cpuid_values + 0xa, cpuid_values + 0xb);

    string brand(reinterpret_cast<char*>(cpuid_values), sizeof(cpuid_values));
    return brand;
}

// I couldn't find a reliable way to get cpu architecture using CPUID, other 2 methods will be used (i will try harder to find them later)

const int cores() {
    uint32_t output[4];
    __cpuid(0x00000001, output[0], output[1], output[2], output[3]);
    return (output[1] >> 16) & 0xFF; // Bits 23-16 of EBX contain the number of cores per processor
}

// I couldn't find a reliable way to get cpu clockspeed using CPUID, other 2 methods will be used (i will try harder to find them later)

const string vendor() {
    uint32_t output[4];

    // Get vendor string from CPUID function 0
    __cpuid(0x00000000, output[0], output[1], output[2], output[3]);

    string vendor;
    vendor.append(reinterpret_cast<char*>(&output[1]), sizeof(uint32_t));
    vendor.append(reinterpret_cast<char*>(&output[3]), sizeof(uint32_t));
    vendor.append(reinterpret_cast<char*>(&output[2]), sizeof(uint32_t));

    return vendor;
}

const string manufacturer() {
    const string s_vendor = vendor();
    auto it = vendor_map.find(s_vendor);
    return it != vendor_map.end() ? it->second : "Unknown";
}

File: ospylib/hardware/arch/windows/processor.py
# Copyright (c) 2022, Bamboooz
# All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import winreg

from ospylib.common.registry import read_registry_value
from ospylib.common.prompt import execute_command

WMIC_CPU_MODEL = 'wmic cpu get name'
WMIC_CPU_CORES = 'wmic cpu get NumberOfLogicalProcessors'
WMIC_CPU_CLOCKSPEED = 'wmic cpu get maxclockspeed'
WMIC_CPU_ARCHITECTURE = 'wmic cpu get Architecture'
WMIC_CPU_VENDOR_ID = 'wmic cpu get manufacturer'

architectures = {
    0: "x86",
    5: "ARM",
    6: "Intel Itanium-based",
    9: "AMD64",
    10: "DEC Alpha",
    12: "ARM64",
    13: "MIPS",
    14: "MIPS64",
    15: "Power PC",
    16: "Power PC64",
    17: "SPARC",
    18: "SPARC64",
    20: "SHx",
    21: "SCx"
}

vendors = {
    "amdisbetter!": "AMD",
    "authenticamd": "AMD",
    "centaurhauls": "IDT WinChip/Centaur",
    "cyrixinstead": "Cyrix, STMicroelectronics, IBM",
    "genuineintel": "Intel",
    "transmetacpu": "Transmeta",
    "genuinetmx86": "Transmeta",
    "geodebynsc": "National Semiconductor",
    "nexgendriven": "NexGen",
    "riseriserise": "Rise",
    "sississis": "SiS",
    "umcumcumc": "UMC",
    "viaviavia": "VIA",
    "vortex86soc": "DM&P Vortex86",
    "shanghai": "Zhaoxin",
    "hygongenuine": "Hygon",
    "genuinerdc": "RDC Semiconductor Co. Ltd.",
    "e2kmachine": "MCST Elbrus",
}


class wmic:
    @staticmethod
    def model() -> str:
        model = execute_command(WMIC_CPU_MODEL, 1).strip()
        return model

    @staticmethod
    def cores() -> int:
        cores = execute_command(WMIC_CPU_CORES, 1).strip()
        return int(cores)

    @staticmethod
    def architecture() -> str:
        arch_value = int(execute_command(WMIC_CPU_ARCHITECTURE, 1).strip())
        architecture = architectures.get(arch_value, None)
        return architecture

    @staticmethod
    def clockspeed() -> int:
        clockspeed = execute_command(WMIC_CPU_CLOCKSPEED, 1).strip()
        return int(clockspeed)

    @staticmethod
    def vendor() -> str:
        vendor = execute_command(WMIC_CPU_VENDOR_ID, 1).strip()
        return vendor

    @staticmethod
    def manufacturer() -> str:
        vendor_id = execute_command(WMIC_CPU_VENDOR_ID, 1).strip()
        manufacturer = vendors.get(vendor_id.lower().replace(' ', ''), None)
        return manufacturer


class registry:
    @staticmethod
    def model() -> str:
        model = read_registry_value(winreg.HKEY_LOCAL_MACHINE, "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0", "ProcessorNameString").strip()
        return model

    @staticmethod
    def cores() -> int:
        cpu_key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, "HARDWARE\\DESCRIPTION\\System\\CentralProcessor")
        num_cpus = winreg.QueryInfoKey(cpu_key)[0]
        winreg.CloseKey(cpu_key)
        return int(num_cpus)

    @staticmethod
    def architecture() -> str:
        architecture = read_registry_value(winreg.HKEY_LOCAL_MACHINE, "SYSTEM\\CurrentControlSet\\Control\\Session Manager\\Environment", "PROCESSOR_ARCHITECTURE").strip()
        return architecture

    @staticmethod
    def clockspeed() -> int:
        clockspeed = read_registry_value(winreg.HKEY_LOCAL_MACHINE, "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0", "~MHz")
        return clockspeed

    @staticmethod
    def vendor() -> str:
        vendor = read_registry_value(winreg.HKEY_LOCAL_MACHINE, "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0", "VendorIdentifier").strip()
        return vendor

    @staticmethod
    def manufacturer() -> str:
        vendor_id = read_registry_value(winreg.HKEY_LOCAL_MACHINE, "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0", "VendorIdentifier").strip()
        manufacturer = vendors.get(vendor_id.lower().replace(' ', ''), None)
        return manufacturer
